documentation/specs/proposed/BuildCheck.md

```diff
@@ -0,0 +1,405 @@
+- [Scope of initial iteration](#scope-of-initial-iteration)
+- [User Experience](#user-experience)
+   * [Running / UX](#running-ux)
+      + [Inbox Analyzers](#inbox-analyzers)
+      + [Live Build](#live-build)
+      + [Binlog Replay mode](#binlog-replay-mode)
+   * [Configuration](#configuration)
+      + [User Configurable Options](#user-configurable-options)
+         - [Enablement](#enablement)
+         - [Severity](#severity)
+         - [Scope of Analysis](#scope-of-analysis)
+      + [Sample configuration](#sample-configuration)
+   * [Analyzers and Rules Identification](#analyzers-and-rules-identification)
+   * [Custom Analyzers Authoring](#custom-analyzers-authoring)
+      + [Implementation](#implementation)
+         - [Analyzer declaration](#analyzer-declaration)
+         - [Rules declaration](#rules-declaration)
+         - [Standardized configuration declaration](#standardized-configuration-declaration)
+         - [Custom configuration declaration](#custom-configuration-declaration)
+         - [Compatibility](#compatibility)
+      + [Testing and Debugging](#testing-and-debugging)
+      + [Packaging](#packaging)
+   * [Acquisition of custom analyzers](#acquisition-of-custom-analyzers)
+
+
+
+# Terminology
+
+* **Analyzer** – unit of execution (single implementing class), can host multiple rules. 
+* **Rule** – Single violation type, with single unique code (`“MSB123: Redefining built-in target”`). 
+* **Report** – Output from Analyzer informing about violating particular rule.
+* **CodeFix** – Violation remedy suggestion. Not yet applicable for MSBuild.
+* **BuildCheck** - Feature name. The infrastructure within MSBuild allowing pluggability and execution of Analyzers and their Rules
+
+
+# North Star / Longer-term vision
+
+MSBuild provides a rich OM exposing the build scripts, data and execution so that various quality checking rules can be authored. Identical OM is exposed from live build execution and via post-build log event sourcing – so that users can choose whether the build analysis will happen as part of the build or as a separate process.
+
+Users are able to tune the behavior of the checks via `.editorconfig` which brings unified and standardized experience with other tooling leveraging `.editorconfig` files.
+
+Powerusers are able to develop, test and publish their custom analyzers easily and contribute them back to community. The local development scenario doesn’t require roundtrip through packaging.
+
+Solid set of in-the-box analyzers is provided by the dotnet team, extended each release, with high quality reports (pointing exact locations of issue, offering clear and actionable explanations, not repetitive for builds with multi-execution or/and multi-importing of a same script in single build context).
+
+The analysis has small impact on build duration with ability to opt-out from analysis altogether which will remove all the performance costs associated with the analysis.
```

Comment:
Opt out or opt in? I feel perf DDRITs will require opt-in . . .

Reply:
So very intitialy (pre-release) - the analyzers will be opt-in.

In some mature state the analyzers will be opt-out.

The big question is when we switch between those two modes - V1 / V2? Current wishfull thinking is V1.

I'll try to mention and discuss this more explicitly in the doc